### PR TITLE
Limit resource usage of message handling

### PR DESCRIPTION
**Pull request by limebell**
Some messages causes receiver to spawn task that requires huge amount of resources. This patch adds `ResourceRegulationOptions` option to regulate maximum spawning of such tasks.

### CODE REVIEW

File: CHANGES.md [hunk 1]
```diff
@@ -57,6 +57,8 @@ To be released.
 ### Added APIs
 
  -  Added `IActonEvaluator` interface.  [[#3082]]
+ -  Added `SwarmOptions.ResourceRegulationOptions` property.  [[#3109]]
```
Comment (longfin) [inline CHANGES.md:60]: How about `SwarmOptions.TaskRegulationOptions`? (or, is there any plan for applying this option other than task counts?)

File: Libplanet.Net/Swarm.MessageHandlers.cs [hunk 2]
```diff
@@ -186,26 +190,39 @@ private void ProcessBlockHeader(Message message)
 
         private async Task TransferTxsAsync(Message message)
         {
-            var getTxsMsg = (GetTxsMsg)message.Content;
-            foreach (TxId txid in getTxsMsg.TxIds)
+            if (!(_transferTxsSemaphore is null) &&
+                !_transferTxsSemaphore.WaitOne(TimeSpan.Zero))
+            {
+                return;
```
Comment (longfin) [inline Libplanet.Net/Swarm.MessageHandlers.cs:196]: Noticing the remote peer about this resource outage might be helpful. (of course, it causes another response type).

Comment (longfin) [inline Libplanet.Net/Swarm.MessageHandlers.cs:196]: Logging also seems to be needed.

File: Libplanet.Net/Swarm.cs [hunk 3]
```diff
@@ -97,6 +98,18 @@ public partial class Swarm<T> : IDisposable
             Transport.ProcessMessageHandler.Register(ProcessMessageHandlerAsync);
             PeerDiscovery = new KademliaProtocol(RoutingTable, Transport, Address);
 
+            // Regulate heavy tasks. Treat negative value as 0.
+            var resourceRegulationOptions = Options.ResourceRegulationOptions;
+            int limit = resourceRegulationOptions.MaxTransferBlocksTaskCount;
+            _transferBlocksSemaphore = limit > 0
+                ? new Semaphore(limit, limit)
+                : null;
```
Comment (longfin) [inline Libplanet.Net/Swarm.cs:106]: Introducing a new static helper method to convert int to semaphore may help for readability.

File: Libplanet.Net.Tests/SwarmTest.cs [hunk 4]
```diff
@@ -1792,6 +1793,129 @@ public async Task LastMessageTimestamp()
...
+                for (int i = 0; i < 5; i++)
+                {
+                    tasks.Add(
+                        transport.SendMessageAsync(
+                            swarm.AsPeer,
+                            content,
+                            TimeSpan.FromMilliseconds(1000),
+                            1,
+                            false,
+                            default));
```
Comment (longfin) [inline Libplanet.Net.Tests/SwarmTest.cs:1834]:
```suggestion
                        Task.Run(async () => await transport.SendMessageAsync(
                            swarm.AsPeer,
                            content,
                            TimeSpan.FromMilliseconds(1000),
                            1,
                            false,
                            default)));
```

`Task.Run()` may be needed especially in xunit test cases (holding SynchronizationContext)

File: Libplanet.Net/NullableSemaphore.cs [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+        public void Dispose()
+        {
+            _sema?.Dispose();
+        }
+
+        public async Task<bool> WaitAsync(TimeSpan timeout, CancellationToken cancellationToken)
+        {
+            if (_sema is { } sema)
+            {
+                return await sema.WaitAsync(timeout, cancellationToken);
```
Comment (longfin) [inline Libplanet.Net/NullableSemaphore.cs:27]:
```suggestion
                return await sema.WaitAsync(timeout, cancellationToken).ConfigureAwait(false);
```

**Comment by limebell** [comment]
@longfin Applied suggestions in f907d77acbcd54af95b5a48f08d97b01a9202a16. PTAL

**Review by longfin — APPROVED** [review]

**Pull request merged**